keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `k8s.io/component-base/kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```go
+import "k8s.io/component-base/kontext"
+
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx).WithName("probe")
+  logger.V(6).Info("Updating ready status", "ready", newStatus)
+  [...]
+  logger.V(4).Info("Ready status updated", "ready", newStatus)
+}
+```
+
+### New logging formats
+
+Introduction of new logging API should be paired with new output format that can properly utilize structured logging. In order to simplify log ingestion, but still maintain similar developer experience we would like to propose two formats. First one for production use and second one for component development.
+
+#### Json format
+
+We would like to propose a JSON as a future default format in Kubernetes for both production and development use. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
+
+```json
+{
+   "ts":"00:15:15.525108",
+   "msg":"Updating ready status",
+   "core/node":{
+      "name":"node-1"
+   },
+   "core/pod":{
+      "name":"nginx-1",
+      "namespace":"default"
+   },
+   "probe":{
+      "ready":false
+   }
+}
+```
+
+### Logging configuration
+
+We would like to introduce new logging configuration options shared by all kubernetes components. `LoggingConfig` structure should be implemented as part of `k8s.io/component-base` options and include common set of flags for logger initialization.
+* `--logging-format` flag should allow to pick between logging interface implementation and allow for backward compatibility. Setting this flag will select particular logger implementation from registry.
+* `-v` flag should allow to change verbosity of logging
+
+Any additional configuration needed by loggers should be provided by flags specific to its implementation. Default loggers should provide a basic customization, but it will not be encouraged to avoid feature bloat. Adding new loggers, changing logger behavior or extending its functionality should be done outside of Kubernetes by implementing logr API and registering it in code.
```

Comment:
can you link to a sample mechanism for registration?  my initial read of this is that any alternative logger still needs to be vendored, but its possible i am missing something...

Reply:
As of now we have not planned to provide any specific mechanism for registration of those logr implementations. So this point is more about signaling the future development area which should be taken into consideration when implementing standard logger configuration and which should be followed when we will gather more input on this.

Reply:
I think that's the correct read, though (in theory) Go's plugins system is usable for this.  Whoever builds the main binaries also builds an assortment of common log libs.  Binaries take a standardized flag saying which log-lib to use and dlopen it very early.  The log libs need to be managed, but maybe not in k/k itself.